### PR TITLE
add better healthcheck

### PR DESCRIPTION
**Pull request by omer9564**

### CODE REVIEW

File: horizon/enforcer/utils/statistics_utils.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        while True:
+            message = await self._messages.get()
+            try:
+                async with self._lock:
+                    self._requests += 1
+                    if message is False:
+                        self._failures += 1
+            finally:
+                self._messages.task_done()
+            await asyncio.sleep(0.5)
```
Comment (asafc) [inline horizon/enforcer/utils/statistics_utils.py:24]: why do you need to sleep here?
won't we await without blocking on the next message in the queue?

**Review by asafc — APPROVED** [review]
there are some small things to fix but i don't want to block you.

File: horizon/enforcer/utils/statistics_utils.py [hunk 2]
```diff
@@ -1,47 +1,77 @@
 import asyncio
 
+from loguru import logger
+
 
 class StatisticsManager:
-    def __init__(self, interval_seconds: int = 2 * 60):
+    def __init__(self, interval_seconds: int = 60, failures_threshold_percentage: float = 10):
```
Comment (asafc) [inline horizon/enforcer/utils/statistics_utils.py:7]: failures_threshold_percentage: float = 0.1

?

File: horizon/enforcer/utils/statistics_utils.py [hunk 3]
```diff
@@ -1,47 +1,77 @@
...
+        await self._messages.join()
+        if self._restarter_task is not None:
+            self._restarter_task.cancel()
+            self._restarter_task = None
+        if self._interval_task is not None:
+            self._interval_task.cancel()
+            self._interval_task = None
 
     def report_success(self) -> None:
+        logger.info("Reporting success")
```
Comment (asafc) [inline horizon/enforcer/utils/statistics_utils.py:60]: this should definitely be `debug`.

File: horizon/enforcer/utils/statistics_utils.py [hunk 4]
```diff
@@ -1,47 +1,77 @@
...
         self._messages.put_nowait(True)
 
     def report_failure(self) -> None:
+        logger.info("Reporting failure")
```
Comment (asafc) [inline horizon/enforcer/utils/statistics_utils.py:64]: this should definitely be `debug`.

File: horizon/enforcer/utils/statistics_utils.py [hunk 5]
```diff
@@ -1,47 +1,77 @@
...
+        current_requests, current_failures = float(self._requests), float(self._failures)
+        if current_requests == 0:
+            return 0.0
         return current_failures / current_requests
+
+    async def status(self) -> bool:
+        rate = await self.current_rate()
+        if rate > self._failures_threshold_percentage:
+            self._had_failure = True
+        return self._had_failure
```
Comment (asafc) [inline horizon/enforcer/utils/statistics_utils.py:77]: not sure if this should be separated into a separate check or not.
i.e:
- currently having failures
- had failures (at least once)

not a blocker, just wondering.

**Review by asafc — COMMENTED** [review]
few nits, looking good.

**Review by asafc — APPROVED** [review]

File: horizon/enforcer/api.py [hunk 6]
```diff
@@ -258,6 +265,24 @@ def init_enforcer_api_router(policy_store: BasePolicyStoreClient = None):
...
+    stats_manager.run()
+
+    @router.get(
+        "/health",
+        status_code=status.HTTP_200_OK,
+        include_in_schema=False
+    )
+    async def health():
+        current_rate = await stats_manager.current_rate()
+        if current_rate > sidecar_config.OPA_CLIENT_FAILURE_THRESHOLD_PERCENTAGE:
```
[the remark this answers is not in the document]
Comment (omer9564) [reply horizon/enforcer/api.py]: changed

File: horizon/config.py [hunk 7]
```diff
@@ -74,6 +74,16 @@ class SidecarConfig(Confi):
...
         description="the timeout for querying OPA for an allow decision, in seconds. 0 means no timeout",
     )
+    OPA_CLIENT_FAILURE_THRESHOLD_PERCENTAGE = confi.float(
+        "OPA_CLIENT_FAILURE_THRESHOLD",
+        10,
+        description="the percentage of failed requests to OPA that will trigger a failure threshold",
+    )
+    OPA_CLIENT_FAILURE_THRESHOLD_INTERVAL = confi.float(
+        "OPA_CLIENT_FAILURE_THRESHOLD_INTERVAL",
+        2,
```
[the remark this answers is not in the document]
Comment (omer9564) [reply horizon/config.py]: changed

**Pull request merged**